File: android/app/src/main/java/io/conduit/network/FrameCodec.kt
package io.conduit.network

import java.io.DataInputStream
import java.io.EOFException
import java.io.InputStream
import java.io.OutputStream

/**
 * Length-prefixed framing: [4-byte big-endian length][payload]. Matches the .NET FrameCodec.
 */
object FrameCodec {
    const val MAX_FRAME_SIZE = 16 * 1024 * 1024

    fun writeFrame(out: OutputStream, payload: ByteArray) {
        require(payload.size <= MAX_FRAME_SIZE) { "Frame too large: ${payload.size}" }
        val len = payload.size
        val header = byteArrayOf(
            (len ushr 24).toByte(),
            (len ushr 16).toByte(),
            (len ushr 8).toByte(),
            len.toByte(),
        )
        synchronized(out) {
            out.write(header)
            out.write(payload)
            out.flush()
        }
    }

    /** Reads one full frame, or returns null on clean end-of-stream. */
    fun readFrame(input: InputStream): ByteArray? {
        val din = if (input is DataInputStream) input else DataInputStream(input)
        val header = ByteArray(4)
        val first = input.read()
        if (first == -1) return null
        header[0] = first.toByte()
        readExact(din, header, 1, 3)

        val len = ((header[0].toInt() and 0xFF) shl 24) or
            ((header[1].toInt() and 0xFF) shl 16) or
            ((header[2].toInt() and 0xFF) shl 8) or
            (header[3].toInt() and 0xFF)
        if (len < 0 || len > MAX_FRAME_SIZE) throw IllegalStateException("Bad frame size $len")
        if (len == 0) return ByteArray(0)

        val payload = ByteArray(len)
        readExact(din, payload, 0, len)
        return payload
    }

    private fun readExact(din: DataInputStream, buf: ByteArray, offset: Int, count: Int) {
        var read = 0
        while (read < count) {
            val n = din.read(buf, offset + read, count - read)
            if (n == -1) throw EOFException("Stream ended mid-frame")
            read += n
        }
    }
}

File: windows/src/Conduit.Core/Networking/FrameCodec.cs
using System.Buffers.Binary;

namespace Conduit.Core.Networking;

/// <summary>
/// Reads/writes length-prefixed frames on a stream: [4-byte big-endian length][payload].
/// This is the low-level framing described in PROTOCOL.md §2.
/// </summary>
public static class FrameCodec
{
    /// <summary>Maximum single frame size (16 MB) — guards against corrupt/hostile length prefixes.</summary>
    public const int MaxFrameSize = 16 * 1024 * 1024;

    public static async Task WriteFrameAsync(Stream stream, byte[] payload, CancellationToken ct)
    {
        if (payload.Length > MaxFrameSize)
            throw new InvalidOperationException($"Frame too large: {payload.Length} bytes");

        var header = new byte[4];
        BinaryPrimitives.WriteUInt32BigEndian(header, (uint)payload.Length);
        await stream.WriteAsync(header, ct).ConfigureAwait(false);
        await stream.WriteAsync(payload, ct).ConfigureAwait(false);
        await stream.FlushAsync(ct).ConfigureAwait(false);
    }

    /// <summary>Reads one full frame. Returns null on clean end-of-stream.</summary>
    public static async Task<byte[]?> ReadFrameAsync(Stream stream, CancellationToken ct)
    {
        var header = await ReadExactAsync(stream, 4, ct).ConfigureAwait(false);
        if (header is null) return null;

        uint len = BinaryPrimitives.ReadUInt32BigEndian(header);
        if (len > MaxFrameSize)
            throw new InvalidDataException($"Declared frame size {len} exceeds max {MaxFrameSize}");
        if (len == 0) return [];

        var payload = await ReadExactAsync(stream, (int)len, ct).ConfigureAwait(false);
        if (payload is null)
            throw new EndOfStreamException("Stream ended mid-frame");
        return payload;
    }

    private static async Task<byte[]?> ReadExactAsync(Stream stream, int count, CancellationToken ct)
    {
        var buffer = new byte[count];
        int read = 0;
        while (read < count)
        {
            int n = await stream.ReadAsync(buffer.AsMemory(read, count - read), ct).ConfigureAwait(false);
            if (n == 0)
                return read == 0 ? null : throw new EndOfStreamException("Stream ended mid-frame");
            read += n;
        }
        return buffer;
    }
}
